File: examples/ts-server/src/presentation/resolvers.ts
/**
 * GraphQL Resolvers
 *
 * Maps GraphQL operations to application layer use cases.
 * Handles result types and converts domain errors to GraphQL responses.
 */

import type { Resolvers } from "@bgql/server";
import { Context, requireAuth, requireRole } from "./context.js";
import {
  User, Post, Comment, UserId, PostId, UserRole,
  PostStatus, UserAnalytics,
} from "../domain/entities.js";
import { DomainError } from "../domain/errors.js";

// ============================================
// Helper: Convert Result to Union Response
// ============================================

type ErrorResponse = { __typename: string; message: string; code: string; [key: string]: unknown };

function resultToUnion<T, E extends DomainError>(
  result: { ok: true; value: T } | { ok: false; error: E },
  successMapper: (value: T) => { __typename: string; [key: string]: unknown }
): { __typename: string; [key: string]: unknown } {
  if (result.ok) {
    return successMapper(result.value);
  }
  return result.error.toGraphQL();
}

// ============================================
// Query Resolvers
// ============================================

const queryResolvers = {
  user: async (
    _parent: unknown,
    args: { id: string },
    ctx: Context
  ) => {
    const result = await ctx.services.userQuery.getUser({ id: UserId(args.id) });
    return resultToUnion(result, (user) => ({
      __typename: "User",
      ...userToGraphQL(user),
    }));
  },

  users: async (
    _parent: unknown,
    args: { filter?: { role?: UserRole; search?: string } },
    ctx: Context
  ) => {
    const users = await ctx.services.userQuery.listUsers({ filter: args.filter });
    return users.map((user) => ({
      __typename: "User",
      ...userToGraphQL(user),
    }));
  },

  post: async (
    _parent: unknown,
    args: { id: string },
    ctx: Context
  ) => {
    const result = await ctx.services.postQuery.getPost({ id: PostId(args.id) });
    return resultToUnion(result, (post) => ({
      __typename: "Post",
      ...postToGraphQL(post),
    }));
  },

  posts: async (
    _parent: unknown,
    args: { filter?: { status?: PostStatus; authorId?: string; search?: string } },
    ctx: Context
  ) => {
    const filter = args.filter ? {
      ...args.filter,
      authorId: args.filter.authorId ? UserId(args.filter.authorId) : undefined,
    } : undefined;

    const posts = await ctx.services.postQuery.listPosts({ filter });
    return posts.map((post) => ({
      __typename: "Post",
      ...postToGraphQL(post),
    }));
  },

  me: async (_parent: unknown, _args: unknown, ctx: Context) => {
    if (!ctx.currentUser) {
      return {
        __typename: "UnauthorizedError",
        message: "Not authenticated",
        code: "UNAUTHORIZED",
        requiredRole: null,
      };
    }
    return {
      __typename: "User",
      ...userToGraphQL(ctx.currentUser),
    };
  },
};

// ============================================
// Mutation Resolvers
// ============================================

const mutationResolvers = {
  createUser: async (
    _parent: unknown,
    args: { input: { name: string; email: string; bio?: string | null; role?: UserRole } },
    ctx: Context
  ) => {
    const result = await ctx.services.userCommand.createUser({
      name: args.input.name,
      email: args.input.email,
      bio: args.input.bio,
      role: args.input.role,
    });
    return resultToUnion(result, (user) => ({
      __typename: "User",
      ...userToGraphQL(user),
    }));
  },

  updateUser: async (
    _parent: unknown,
    args: { id: string; input: { name?: string; email?: string; bio?: string | null; avatarUrl?: string | null } },
    ctx: Context
  ) => {
    const result = await ctx.services.userCommand.updateUser({
      id: UserId(args.id),
      ...args.input,
    });
    return resultToUnion(result, (user) => ({
      __typename: "User",
      ...userToGraphQL(user),
    }));
  },

  deleteUser: async (
    _parent: unknown,
    args: { id: string },
    ctx: Context
  ) => {
    const currentUser = requireAuth(ctx);
    const result = await ctx.services.userCommand.deleteUser({
      id: UserId(args.id),
      requesterId: currentUser.id,
      requesterRole: currentUser.role,
    });

    if (result.ok) {
      return { __typename: "DeleteSuccess", success: true };
    }
    return result.error.toGraphQL();
  },

  createPost: async (
    _parent: unknown,
    args: { input: { title: string; content: string; status?: PostStatus } },
    ctx: Context
  ) => {
    const currentUser = requireAuth(ctx);
    const result = await ctx.services.postCommand.createPost({
      title: args.input.title,
      content: args.input.content,
      authorId: currentUser.id,
      status: args.input.status,
    });
    return resultToUnion(result, (post) => ({
      __typename: "Post",
      ...postToGraphQL(post),
    }));
  },

  updatePost: async (
    _parent: unknown,
    args: { id: string; input: { title?: string; content?: string; status?: PostStatus } },
    ctx: Context
  ) => {
    const currentUser = requireAuth(ctx);
    const result = await ctx.services.postCommand.updatePost({
      id: PostId(args.id),
      requesterId: currentUser.id,
      requesterRole: currentUser.role,
      ...args.input,
    });
    return resultToUnion(result, (post) => ({
      __typename: "Post",
      ...postToGraphQL(post),
    }));
  },

  deletePost: async (
    _parent: unknown,
    args: { id: string },
    ctx: Context
  ) => {
    const currentUser = requireAuth(ctx);
    const result = await ctx.services.postCommand.deletePost({
      id: PostId(args.id),
      requesterId: currentUser.id,
      requesterRole: currentUser.role,
    });

    if (result.ok) {
      return { __typename: "DeleteSuccess", success: true };
    }
    return result.error.toGraphQL();
  },

  publishPost: async (
    _parent: unknown,
    args: { id: string },
    ctx: Context
  ) => {
    const currentUser = requireAuth(ctx);
    const result = await ctx.services.postCommand.publishPost({
      id: PostId(args.id),
      requesterId: currentUser.id,
      requesterRole: currentUser.role,
    });
    return resultToUnion(result, (post) => ({
      __typename: "Post",
      ...postToGraphQL(post),
    }));
  },
};

// ============================================
// Type Resolvers (for nested fields)
// ============================================

const userTypeResolvers = {
  posts: async (parent: { id: string }, _args: unknown, ctx: Context) => {
    const posts = await ctx.loaders.postsByAuthor.load(UserId(parent.id));
    return posts.map((post) => ({
      __typename: "Post",
      ...postToGraphQL(post),
    }));
  },

  analytics: async (parent: { id: string }, _args: unknown, ctx: Context): Promise<UserAnalytics> => {
    return ctx.services.userQuery.getUserAnalytics(UserId(parent.id));
  },
};

const postTypeResolvers = {
  author: async (parent: { authorId: string }, _args: unknown, ctx: Context) => {
    const user = await ctx.loaders.user.load(UserId(parent.authorId));
    if (!user) {
      return null;
    }
    return {
      __typename: "User",
      ...userToGraphQL(user),
    };
  },

  comments: async (parent: { id: string }, _args: unknown, ctx: Context) => {
    const comments = await ctx.loaders.commentsByPost.load(PostId(parent.id));
    return comments.map((comment) => ({
      __typename: "Comment",
      ...commentToGraphQL(comment),
    }));
  },
};

const commentTypeResolvers = {
  author: async (parent: { authorId: string }, _args: unknown, ctx: Context) => {
    const user = await ctx.loaders.user.load(UserId(parent.authorId));
    if (!user) {
      return null;
    }
    return {
      __typename: "User",
      ...userToGraphQL(user),
    };
  },

  post: async (parent: { postId: string }, _args: unknown, ctx: Context) => {
    const post = await ctx.loaders.post.load(PostId(parent.postId));
    if (!post) {
      return null;
    }
    return {
      __typename: "Post",
      ...postToGraphQL(post),
    };
  },
};

// ============================================
// Domain to GraphQL Mappers
// ============================================

function userToGraphQL(user: User) {
  return {
    id: user.id,
    name: user.name,
    email: user.email,
    bio: user.bio,
    avatarUrl: user.avatarUrl,
    role: user.role,
    createdAt: user.createdAt.toISOString(),
    updatedAt: user.updatedAt?.toISOString() ?? null,
  };
}

function postToGraphQL(post: Post) {
  return {
    id: post.id,
    title: post.title,
    content: post.content,
    authorId: post.authorId,
    status: post.status,
    publishedAt: post.publishedAt?.toISOString() ?? null,
    createdAt: post.createdAt.toISOString(),
  };
}

function commentToGraphQL(comment: Comment) {
  return {
    id: comment.id,
    content: comment.content,
    authorId: comment.authorId,
    postId: comment.postId,
    createdAt: comment.createdAt.toISOString(),
  };
}

// ============================================
// Export Combined Resolvers
// ============================================

export const resolvers: Resolvers<Context> = {
  Query: queryResolvers,
  Mutation: mutationResolvers,
  User: userTypeResolvers,
  Post: postTypeResolvers,
  Comment: commentTypeResolvers,
};

File: examples/rust-server/src/presentation/resolvers.rs
//! GraphQL resolvers - Type-safe implementation using generated traits.

use crate::application::*;
use crate::generated::*;
use crate::infrastructure::*;
use async_trait::async_trait;
use bgql_sdk::server::Context;
use bgql_sdk::SdkResult;
use serde_json::json;
use std::sync::Arc;

/// Application context containing all services.
pub struct AppContext {
    pub user_service: UserService,
    pub post_service: PostService,
    pub comment_service: CommentService,
}

impl AppContext {
    pub fn new(
        user_repo: Arc<dyn UserRepository>,
        post_repo: Arc<dyn PostRepository>,
        comment_repo: Arc<dyn CommentRepository>,
    ) -> Self {
        Self {
            user_service: UserService::new(user_repo.clone(), post_repo.clone()),
            post_service: PostService::new(post_repo.clone(), user_repo.clone()),
            comment_service: CommentService::new(comment_repo, post_repo),
        }
    }
}

/// Type-safe Query resolvers implementation.
pub struct AppQueryResolvers {
    ctx: Arc<AppContext>,
}

impl AppQueryResolvers {
    pub fn new(ctx: Arc<AppContext>) -> Self {
        Self { ctx }
    }
}

#[async_trait]
impl QueryResolvers for AppQueryResolvers {
    async fn user(&self, _ctx: &Context, args: UserArgs) -> SdkResult<serde_json::Value> {
        let id = crate::domain::UserId::new(&args.id.0);

        match self.ctx.user_service.get_user_with_posts(&id).await {
            Ok((user, posts)) => Ok(json!({
                "__typename": "User",
                "id": user.id.0,
                "name": user.name,
                "email": user.email,
                "bio": user.bio,
                "createdAt": user.created_at.to_rfc3339(),
                "posts": posts.iter().map(|p| json!({
                    "id": p.id.0,
                    "title": p.title,
                    "status": p.status.to_string()
                })).collect::<Vec<_>>(),
                "postsCount": posts.len()
            })),
            Err(crate::domain::DomainError::UserNotFound(id)) => Ok(json!({
                "__typename": "NotFoundError",
                "message": format!("User '{}' not found", id),
                "code": "NOT_FOUND",
                "resourceType": "User",
                "resourceId": id
            })),
            Err(e) => Ok(json!({
                "__typename": "ValidationError",
                "message": e.to_string(),
                "code": "ERROR"
            })),
        }
    }

    async fn users(&self, _ctx: &Context, args: UsersArgs) -> SdkResult<serde_json::Value> {
        let pagination = args
            .pagination
            .map(|p| crate::domain::Pagination {
                first: p.first.map(|n| n as usize),
                after: p.after,
            })
            .unwrap_or_default();

        let conn = self.ctx.user_service.list_users(pagination).await;

        Ok(json!({
            "edges": conn.edges.iter().map(|e| json!({
                "cursor": e.cursor,
                "node": {
                    "id": e.node.id.0,
                    "name": e.node.name,
                    "email": e.node.email,
                    "bio": e.node.bio,
                    "createdAt": e.node.created_at.to_rfc3339()
                }
            })).collect::<Vec<_>>(),
            "pageInfo": {
                "hasNextPage": conn.page_info.has_next_page,
                "hasPreviousPage": conn.page_info.has_previous_page,
                "startCursor": conn.page_info.start_cursor,
                "endCursor": conn.page_info.end_cursor
            },
            "totalCount": conn.total_count
        }))
    }

    async fn post(&self, _ctx: &Context, args: PostArgs) -> SdkResult<serde_json::Value> {
        let id = crate::domain::PostId::new(&args.id.0);

        match self.ctx.post_service.get_post_with_author(&id).await {
            Ok((post, author)) => Ok(json!({
                "__typename": "Post",
                "id": post.id.0,
                "title": post.title,
                "content": post.content,
                "status": post.status.to_string(),
                "authorId": post.author_id.0,
                "author": author.map(|a| json!({
                    "id": a.id.0,
                    "name": a.name,
                    "email": a.email
                })),
                "createdAt": post.created_at.to_rfc3339(),
                "updatedAt": post.updated_at.map(|t| t.to_rfc3339())
            })),
            Err(crate::domain::DomainError::PostNotFound(id)) => Ok(json!({
                "__typename": "NotFoundError",
                "message": format!("Post '{}' not found", id),
                "code": "NOT_FOUND",
                "resourceType": "Post",
                "resourceId": id
            })),
            Err(e) => Ok(json!({
                "__typename": "ValidationError",
                "message": e.to_string(),
                "code": "ERROR"
            })),
        }
    }

    async fn posts(&self, _ctx: &Context, args: PostsArgs) -> SdkResult<serde_json::Value> {
        let pagination = args
            .pagination
            .map(|p| crate::domain::Pagination {
                first: p.first.map(|n| n as usize),
                after: p.after,
            })
            .unwrap_or_default();

        let filter = args
            .filter
            .map(|f| crate::domain::PostFilter {
                status: f
                    .status
                    .and_then(|s| crate::domain::PostStatus::from_str(&s.to_string())),
                author_id: f.author_id.map(|id| crate::domain::UserId::new(&id.0)),
            })
            .unwrap_or_default();

        let conn = self.ctx.post_service.list_posts(filter, pagination).await;
        let author_ids: Vec<_> = conn
            .edges
            .iter()
            .map(|e| e.node.author_id.clone())
            .collect();
        let authors = self.ctx.user_service.get_users_by_ids(&author_ids).await;

        Ok(json!({
            "edges": conn.edges.iter().map(|e| {
                let author = authors.get(&e.node.author_id);
                json!({
                    "cursor": e.cursor,
                    "node": {
                        "id": e.node.id.0,
                        "title": e.node.title,
                        "content": e.node.content,
                        "status": e.node.status.to_string(),
                        "authorId": e.node.author_id.0,
                        "author": author.map(|a| json!({"id": a.id.0, "name": a.name})),
                        "createdAt": e.node.created_at.to_rfc3339()
                    }
                })
            }).collect::<Vec<_>>(),
            "pageInfo": {
                "hasNextPage": conn.page_info.has_next_page,
                "hasPreviousPage": conn.page_info.has_previous_page,
                "startCursor": conn.page_info.start_cursor,
                "endCursor": conn.page_info.end_cursor
            },
            "totalCount": conn.total_count
        }))
    }
}

/// Type-safe Mutation resolvers implementation.
pub struct AppMutationResolvers {
    ctx: Arc<AppContext>,
}

impl AppMutationResolvers {
    pub fn new(ctx: Arc<AppContext>) -> Self {
        Self { ctx }
    }
}

#[async_trait]
impl MutationResolvers for AppMutationResolvers {
    async fn create_post(
        &self,
        _ctx: &Context,
        args: CreatePostArgs,
    ) -> SdkResult<serde_json::Value> {
        let input = crate::application::CreatePostInput {
            author_id: crate::domain::UserId::new("user_1"), // TODO: from auth context
            title: args.input.title,
            content: args.input.content,
            status: args
                .input
                .status
                .and_then(|s| crate::domain::PostStatus::from_str(&s.to_string())),
        };

        match self.ctx.post_service.create_post(input).await {
            Ok(post) => Ok(json!({
                "__typename": "Post",
                "id": post.id.0,
                "title": post.title,
                "content": post.content,
                "status": post.status.to_string(),
                "createdAt": post.created_at.to_rfc3339()
            })),
            Err(crate::domain::DomainError::ValidationError { field, message }) => Ok(json!({
                "__typename": "ValidationError",
                "message": message,
                "code": "VALIDATION_ERROR",
                "field": field
            })),
            Err(e) => Ok(json!({
                "__typename": "ValidationError",
                "message": e.to_string(),
                "code": "ERROR"
            })),
        }
    }

    async fn update_post(
        &self,
        _ctx: &Context,
        args: UpdatePostArgs,
    ) -> SdkResult<serde_json::Value> {
        let id = crate::domain::PostId::new(&args.id.0);
        let input = crate::application::UpdatePostInput {
            title: args.input.title,
            content: args.input.content,
            status: args
                .input
                .status
                .and_then(|s| crate::domain::PostStatus::from_str(&s.to_string())),
        };

        match self.ctx.post_service.update_post(&id, input).await {
            Ok(post) => Ok(json!({
                "__typename": "Post",
                "id": post.id.0,
                "title": post.title,
                "status": post.status.to_string()
            })),
            Err(crate::domain::DomainError::PostNotFound(id)) => Ok(json!({
                "__typename": "NotFoundError",
                "message": format!("Post '{}' not found", id),
                "code": "NOT_FOUND"
            })),
            Err(crate::domain::DomainError::ValidationError { field, message }) => Ok(json!({
                "__typename": "ValidationError",
                "message": message,
                "field": field
            })),
            Err(e) => Ok(json!({
                "__typename": "ValidationError",
                "message": e.to_string()
            })),
        }
    }

    async fn publish_post(
        &self,
        _ctx: &Context,
        args: PublishPostArgs,
    ) -> SdkResult<serde_json::Value> {
        let id = crate::domain::PostId::new(&args.id.0);

        match self.ctx.post_service.publish_post(&id).await {
            Ok(post) => Ok(json!({
                "__typename": "Post",
                "id": post.id.0,
                "title": post.title,
                "status": post.status.to_string()
            })),
            Err(crate::domain::DomainError::PostNotFound(id)) => Ok(json!({
                "__typename": "NotFoundError",
                "message": format!("Post '{}' not found", id),
                "code": "NOT_FOUND"
            })),
            Err(e) => Ok(json!({
                "__typename": "ValidationError",
                "message": e.to_string()
            })),
        }
    }

    async fn delete_post(
        &self,
        _ctx: &Context,
        args: DeletePostArgs,
    ) -> SdkResult<serde_json::Value> {
        let id = crate::domain::PostId::new(&args.id.0);

        match self.ctx.post_service.delete_post(&id).await {
            Ok(()) => Ok(json!({
                "__typename": "DeleteSuccess",
                "success": true,
                "deletedId": args.id.0
            })),
            Err(crate::domain::DomainError::PostNotFound(id)) => Ok(json!({
                "__typename": "NotFoundError",
                "message": format!("Post '{}' not found", id),
                "code": "NOT_FOUND"
            })),
            Err(e) => Ok(json!({
                "__typename": "ValidationError",
                "message": e.to_string()
            })),
        }
    }

    async fn create_comment(
        &self,
        _ctx: &Context,
        args: CreateCommentArgs,
    ) -> SdkResult<serde_json::Value> {
        let input = crate::application::CreateCommentInput {
            post_id: crate::domain::PostId::new(&args.input.post_id.0),
            author_id: crate::domain::UserId::new("user_1"), // TODO: from auth context
            content: args.input.content,
        };

        match self.ctx.comment_service.create_comment(input).await {
            Ok(comment) => Ok(json!({
                "__typename": "Comment",
                "id": comment.id.0,
                "content": comment.content,
                "postId": comment.post_id.0,
                "createdAt": comment.created_at.to_rfc3339()
            })),
            Err(crate::domain::DomainError::PostNotFound(id)) => Ok(json!({
                "__typename": "NotFoundError",
                "message": format!("Post '{}' not found", id),
                "code": "NOT_FOUND"
            })),
            Err(crate::domain::DomainError::ValidationError { field, message }) => Ok(json!({
                "__typename": "ValidationError",
                "message": message,
                "field": field
            })),
            Err(e) => Ok(json!({
                "__typename": "ValidationError",
                "message": e.to_string()
            })),
        }
    }
}
